cabal-install/bootstrap.py

```diff
@@ -0,0 +1,436 @@
+
+        deps = [] # type: List[BootstrapDep]
+        if 'components' in unit:
+            for comp_name, comp in unit['components'].items():
+                deps += unit_ids_deps(comp['depends'])
+        if 'depends' in unit:
+            deps += unit_ids_deps(unit['depends'])
+
+        return deps
+
+    deps = remove_duplicates(unit_deps(target_unit) + [unit_to_bootstrap_dep(target_unit)])
+
+    # Find all of the package revisions and cabal file SHAs.
+    pkgs_needing_rev_info = set((dep.package, dep.version)
+                                for dep in deps
+                                if dep.source == PackageSource.HACKAGE) # type: Set[PackageSpec]
+    rev_info = find_pkg_revisions(HACKAGE_TARBALL, pkgs_needing_rev_info)
+    def update_rev_info(dep: BootstrapDep) -> BootstrapDep:
+        if dep.source == PackageSource.HACKAGE:
+            info = rev_info.get((dep.package, dep.version))
+            assert info is not None
+            cabal_sha256, revision = info
+            dep = dep._replace(cabal_sha256=cabal_sha256, revision=revision)
+            assert dep.revision is not None
+            assert dep.cabal_sha256 is not None
+
+        return dep
+
+    return [update_rev_info(dep) for dep in deps]
+
+a = TypeVar('a')
+
+def remove_duplicates(xs: List[a]) -> List[a]:
+    # it's easier to build lists and remove duplicates later than
+    # to implement an order-preserving set.
+    out = [] # type: List[a]
+    for x in xs:
+        if x not in out:
+            out.append(x)
+
+    return out
+
+def bootstrap(deps: List[BootstrapDep], ghc: Compiler) -> None:
+    if not PKG_DB.exists():
+        print(f'Creating package database {PKG_DB}')
+        PKG_DB.parent.mkdir(parents=True, exist_ok=True)
+        subprocess.run([ghc.ghc_pkg_path, 'init', PKG_DB])
+
+    for dep in deps:
+        install_dep(dep, ghc)
+
+def main() -> None:
+    import argparse
+    parser = argparse.ArgumentParser(
+        description="bootstrapping utility for cabal-install.",
+        epilog = USAGE,
+        formatter_class = argparse.RawDescriptionHelpFormatter)
+    parser.add_argument('--extract-plan', action='store_true',
+                        help='generate bootstrap-deps.json from plan.json')
+    parser.add_argument('-d', '--deps', type=Path, default='bootstrap-deps.json',
+                        help='bootstrap dependency file')
+    parser.add_argument('-w', '--with-compiler', type=Path,
+                        help='path to GHC')
+    args = parser.parse_args()
+
+    # Find compiler
+    if args.with_compiler is None:
+        path = shutil.which('ghc')
+        if path is None:
+            raise ValueError("Couldn't find ghc in PATH")
+        ghc = Compiler(Path(path))
+    else:
+        ghc = Compiler(args.with_compiler)
+
+    print(f'Bootstrapping cabal-install with GHC {ghc.version} at {ghc.ghc_path}...')
+
+    if args.extract_plan:
+        subprocess.run(['cabal', 'v2-configure', '-w', ghc.ghc_path])
```

Comment:
'v2-build', '--dry'

Configure creates / modifies `cabal.project.local` - that's not great.

Also we probably should use `cabal.project.release` so dev stuff don't accidentally affect buildplans. Or even own project file (if we want to use `Cabal` bundled with GHC).